presto-hive/src/main/java/io/prestosql/plugin/hive/MsckProcedure.java

```diff
@@ -0,0 +1,213 @@
+
+        Table table = metastore.getTable(schemaName, tableName).orElse(null);
+        if (table == null) {
+            return;
+        }
+
+        Path defaultLocation = new Path(table.getStorage().getLocation());
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(context, defaultLocation);
+
+            List<String> partitionsInMetastore = metastore.getPartitionNames(schemaName, tableName).orElse(emptyList());
+            List<String> partitionsInFileSystem = listDirectory(fileSystem, fileSystem.getFileStatus(defaultLocation), table.getPartitionColumns(), table.getPartitionColumns().size()).stream()
+                    .map(fileStatus -> defaultLocation.toUri().relativize(fileStatus.getPath().toUri()).getPath())
+                    .collect(toList());
+
+            // partitions in metastore but not in file system
+            List<String> partitionsToDrop = difference(partitionsInMetastore, partitionsInFileSystem);
+            // partitions in file system but not in metastore
+            List<String> partitionsToAdd = difference(partitionsInFileSystem, partitionsInMetastore);
+
+            if (add) {
+                addPartitions(metastore, session, table, defaultLocation, partitionsToAdd);
+            }
+            if (drop) {
+                dropPartitions(metastore, session, table, partitionsToDrop);
+            }
+            metastore.commit();
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private List<FileStatus> listDirectory(FileSystem fileSystem, FileStatus current, List<Column> partitionColumns, int depth)
+    {
+        if (depth == 0) {
+            return ImmutableList.of(current);
+        }
+
+        try {
+            return Stream.of(fileSystem.listStatus(current.getPath()))
+                    .filter(fileStatus -> validatePath(fileSystem, fileStatus, partitionColumns.get(partitionColumns.size() - depth)))
+                    .flatMap(directory -> listDirectory(fileSystem, directory, partitionColumns, depth - 1).stream())
+                    .collect(toList());
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private boolean validatePath(FileSystem fileSystem, FileStatus file, Column column)
+    {
+        try {
+            Path path = file.getPath();
+            String prefix = column.getName() + '=';
+            return fileSystem.isDirectory(path) && path.getName().startsWith(prefix);
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    // calculate relative complement of set b with respect to set a
+    private List<String> difference(List<String> a, List<String> b)
+    {
+        Set<String> diff = Sets.newHashSet(a);
+        diff.removeAll(b);
+        return diff.stream().collect(toList());
```

Comment:
`com.google.common.collect.Sets.difference(new HashSet<>(a), new HashSet<>(b))`

(you can change return type of this method to `Set`)